### PR TITLE
[#828] Validate acceptors, connectors and console exposed with ingress mode

### PR DESCRIPTION
**Pull request by brusdev**

### CODE REVIEW

File: api/v1beta1/activemqartemis_types.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ type ActiveMQArtemisSpec struct {
 	// Optional list of environment variables to apply to the container(s), not exclusive
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Environment Variables"
 	Env []corev1.EnvVar `json:"env,omitempty"`
-	// The ingress domain to expose the application. By default, on Kubernetes it is apps.artemiscloud.io and on OpenShift it is the Ingress Controller domain.
+	// The ingress domain to expose the application. It is required for the acceptors, connectors and console exposed with the ingress mode that do not specify the ingress host.
```
[the remark this answers is not in the document]
Comment (gtully) [reply api/v1beta1/activemqartemis_types.go:68]: ahh, is only used when there is no IngressHost.
maybe we need to call out in the comment - 
// the default ingress domain. It is required when any acceptor, connector or console uses the ingress mode and does not specify an IngressHost.

File: api/v1beta1/activemqartemis_types.go [hunk 2]
```diff
@@ -639,7 +639,7 @@ type ConnectorType struct {
 	// Provider used for the truststore; "SUN", "SunJCE", etc. Default in broker is null
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="TrustStore Provider",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
 	TrustStoreProvider string `json:"trustStoreProvider,omitempty"`
-	// Host for Ingress and Route resources of the acceptor. It supports the following variables: $(CR_NAME), $(CR_NAMESPACE), $(BROKER_ORDINAL), $(ITEM_NAME), $(RES_NAME) and $(INGRESS_DOMAIN). Default is $(CR_NAME)-$(ITEM_NAME)-$(BROKER_ORDINAL)-svc-$(RES_TYPE)-$(CR_NAMESPACE).$(INGRESS_DOMAIN)
+	// Host for Ingress and Route resources of the acceptor. It supports the following variables: $(CR_NAME), $(CR_NAMESPACE), $(BROKER_ORDINAL), $(ITEM_NAME), $(RES_NAME) and $(INGRESS_DOMAIN). . It is required for the connectors exposed with the ingress mode when the ingress domain is not specified.
```
Comment (gtully) [inline api/v1beta1/activemqartemis_types.go:642]: extra dot!

File: controllers/activemqartemis_controller.go [hunk 3]
```diff
@@ -333,12 +333,47 @@ func validateExposeModes(customResource *brokerv1beta1.ActiveMQArtemis) (*metav1
...
 	}
 
+	for _, acceptor := range customResource.Spec.Acceptors {
+		if acceptor.Expose && (acceptor.ExposeMode != nil && *acceptor.ExposeMode == brokerv1beta1.ExposeModes.Ingress || !isOpenshift) &&
+			customResource.Spec.IngressDomain == "" && acceptor.IngressHost == "" {
+			return &metav1.Condition{
+				Type:    brokerv1beta1.ValidConditionType,
+				Status:  metav1.ConditionFalse,
+				Reason:  brokerv1beta1.ValidConditionFailedInvalidIngressSettings,
+				Message: fmt.Sprintf(".Spec.Acceptors %q has invalid ingress settings, no ingressDomain and ingressHost specified", acceptor.Name),
```
Comment (gtully) [inline controllers/activemqartemis_controller.go:349]: ... ingressHost unspecified and no Spec.IngressDomain default domain provided"

The ingressHost is part of the acceptor, but the default domain is part of the Spec. I think it is good to point that out.

**Review by gtully — COMMENTED** [review]
nice, a good addition.

File: controllers/activemqartemis_controller.go [hunk 4]
```diff
@@ -333,12 +333,47 @@ func validateExposeModes(customResource *brokerv1beta1.ActiveMQArtemis) (*metav1
...
+		customResource.Spec.IngressDomain == "" && console.IngressHost == "" {
+		return &metav1.Condition{
+			Type:    brokerv1beta1.ValidConditionType,
+			Status:  metav1.ConditionFalse,
+			Reason:  brokerv1beta1.ValidConditionFailedInvalidIngressSettings,
+			Message: ".Spec.Console has invalid ingress settings, IngressHost unspecified and no Spec.IngressDomain default domain provided",
+		}, false
+	}
+
 	return nil, false
```
Comment (gaohoward) [inline controllers/activemqartemis_controller.go:377]: What if console/acceptor/connector.Expose = false and user specified the **ExposeMode** ? should be valid such conditions?

Comment (brusdev) [reply controllers/activemqartemis_controller.go:377]: I'll add a note in the field documentation to pointed that out.

Comment (gaohoward) [reply controllers/activemqartemis_controller.go:377]: ok.

**Review by gaohoward — APPROVED** [review]

**Pull request merged**